### DISCUSSION TITLE
this project support deepspeed pipeline

### DISCUSSION THREAD

**Discussion by mobilejammer** [bghira/SimpleTuner#1248] [General]
when I creat a config json file to deepspeed, the memory out of also. and seem no part the model

**Comment by bghira** [comment]
please try and complete your thoughts before submitting issue reports or discussions. this doesnt contain information others would need to help you. i mentioned before that you need to at the least submit a config file and the steps taken to lead to the problem. read your own discussion post back; how do you expect anyone to help?

**Comment by mobilejammer** [comment]
I'm terribly sorry that I didn't provide enough information according to your previous reminder to enable you to help me with debugging. I'm training the Flux model with all parameters following the "distributed.md" and "sd3.md". After configuring the DeepSpeed in "accelerate config" to be zero3, I found that the model didn't seem to run in shards. 
![63f8103af03f6bd5870f35df4e02918](https://github.com/user-attachments/assets/7c57be32-5937-44fd-8a3d-6e2758d7facd)
cuda out of memory. When I used Fully Sharded Data Parallel (FSDP), it would run in shards, but it got stuck after the training was completed.
![6314173da0f11f5cb222d867780682e](https://github.com/user-attachments/assets/3af335a4-91ad-4792-a6aa-55cf284dc720)
My graphics card is an NVIDIA GeForce RTX 4090. My Flux FSDP configuration is as follows (the content of the file "~/.cache/huggingface/accelerate/default_config.yaml"):
compute_environment: LOCAL_MACHINE
debug: false
distributed_type: FSDP
downcast_bf16: 'no'
dynamo_config:
dynamo_backend: INDUCTOR
dynamo_mode: default
dynamo_use_dynamic: false
dynamo_use_fullgraph: false
enable_cpu_affinity: false
fsdp_config:
fsdp_activation_checkpointing: true
fsdp_auto_wrap_policy: TRANSFORMER_BASED_WRAP
fsdp_backward_prefetch: BACKWARD_PRE
fsdp_cpu_ram_efficient_loading: true
fsdp_forward_prefetch: true
fsdp_offload_params: true
fsdp_sharding_strategy: FULL_SHARD
fsdp_state_dict_type: SHARDED_STATE_DICT
fsdp_sync_module_states: true
fsdp_transformer_layer_cls_to_wrap: FluxTransformerBlock,FluxSingleTransformerBlock
fsdp_use_orig_params: true
machine_rank: 0
main_training_function: main
mixed_precision: bf16
num_machines: 1
num_processes: 8
rdzv_backend: static
same_network: true
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false
When I tested DeepSpeed Zero3 with multiple machines and multiple GPUs, I still got the "cuda out of memory" error. So I feel that Zero3 doesn't seem to play the real role of splitting large models in this project.

**Comment by mobilejammer** [comment]
when I use fsdp config training:
0:14,190 [INFO] Training has completed.
 -> global_step = 10, max_train_steps = 10, epoch = 1, num_train_epochs = 1
2024-12-28 21:00:14,191 [INFO] Exiting training loop. Beginning model unwind at epoch 1, step 10
192-168-1-86:2006783:2007685 [4] NCCL INFO [Service thread] Connection closed by localRank 4
192-168-1-86:2006784:2007683 [5] NCCL INFO [Service thread] Connection closed by localRank 5
192-168-1-86:2006786:2007682 [7] NCCL INFO [Service thread] Connection closed by localRank 7
192-168-1-86:2006785:2007684 [6] NCCL INFO [Service thread] Connection closed by localRank 6
192-168-1-86:2006781:2007690 [2] NCCL INFO [Service thread] Connection closed by localRank 2
192-168-1-86:2006780:2007687 [1] NCCL INFO [Service thread] Connection closed by localRank 1
192-168-1-86:2006782:2007686 [3] NCCL INFO [Service thread] Connection closed by localRank 3
192-168-1-86:2006786:2013858 [7] NCCL INFO comm 0x5580d734b5f0 rank 7 nranks 8 cudaDev 7 busId e1000 - Abort COMPLETE
192-168-1-86:2006783:2013852 [4] NCCL INFO comm 0x55ddfc00dd00 rank 4 nranks 8 cudaDev 4 busId 81000 - Abort COMPLETE
192-168-1-86:2006784:2013854 [5] NCCL INFO comm 0x5585191f4b60 rank 5 nranks 8 cudaDev 5 busId a1000 - Abort COMPLETE
192-168-1-86:2006785:2013856 [6] NCCL INFO comm 0x55763c6d2370 rank 6 nranks 8 cudaDev 6 busId c1000 - Abort COMPLETE
192-168-1-86:2006782:2013857 [3] NCCL INFO comm 0x561adcdb1580 rank 3 nranks 8 cudaDev 3 busId 61000 - Abort COMPLETE
192-168-1-86:2006781:2013853 [2] NCCL INFO comm 0x5639c2994ed0 rank 2 nranks 8 cudaDev 2 busId 41000 - Abort COMPLETE
192-168-1-86:2006780:2013855 [1] NCCL INFO comm 0x5558289100c0 rank 1 nranks 8 cudaDev 1 busId 23000 - Abort COMPLETE


when I enter ctrl+c to force quite

^CW1228 21:03:38.744000 2006575 torch/distributed/elastic/agent/server/api.py:704] Received Signals.SIGINT death signal, shutting down workers
W1228 21:03:38.747000 2006575 torch/distributed/elastic/multiprocessing/api.py:897] Sending process 2006779 closing signal SIGINT
^CW1228 21:03:39.344000 2006575 torch/distributed/elastic/multiprocessing/api.py:897] Sending process 2006779 closing signal SIGTERM
^CTraceback (most recent call last):
  File "/home/ubuntu/SimpleTuner/.venv/lib/python3.10/site-packages/torch/distributed/elastic/agent/server/api.py", line 696, in run
    result = self._invoke_run(role)
  File "/home/ubuntu/SimpleTuner/.venv/lib/python3.10/site-packages/torch/distributed/elastic/agent/server/api.py", line 855, in _invoke_run
    time.sleep(monitor_interval)
  File "/home/ubuntu/SimpleTuner/.venv/lib/python3.10/site-packages/torch/distributed/elastic/multiprocessing/api.py", line 84, in _terminate_process_handler
    raise SignalException(f"Process {os.getpid()} got signal: {sigval}", sigval=sigval)
torch.distributed.elastic.multiprocessing.api.SignalException: Process 2006575 got signal: 2

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ubuntu/SimpleTuner/.venv/lib/python3.10/site-packages/torch/distributed/elastic/agent/server/api.py", line 705, in run
    self._shutdown(e.sigval)
  File "/home/ubuntu/SimpleTuner/.venv/lib/python3.10/site-packages/torch/distributed/elastic/agent/server/local_elastic_agent.py", line 365, in _shutdown
    self._pcontext.close(death_sig)
  File "/home/ubuntu/SimpleTuner/.venv/lib/python3.10/site-packages/torch/distributed/elastic/multiprocessing/api.py", line 572, in close
    self._close(death_sig=death_sig, timeout=timeout)
  File "/home/ubuntu/SimpleTuner/.venv/lib/python3.10/site-packages/torch/distributed/elastic/multiprocessing/api.py", line 909, in _close
    handler.proc.wait(time_to_wait)
  File "/usr/lib/python3.10/subprocess.py", line 1209, in wait
    return self._wait(timeout=timeout)
  File "/usr/lib/python3.10/subprocess.py", line 1953, in _wait
    time.sleep(delay)
  File "/home/ubuntu/SimpleTuner/.venv/lib/python3.10/site-packages/torch/distributed/elastic/multiprocessing/api.py", line 84, in _terminate_process_handler
    raise SignalException(f"Process {os.getpid()} got signal: {sigval}", sigval=sigval)
torch.distributed.elastic.multiprocessing.api.SignalException: Process 2006575 got signal: 2

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ubuntu/SimpleTuner/.venv/bin/accelerate", line 8, in <module>
    sys.exit(main())
  File "/home/ubuntu/SimpleTuner/.venv/lib/python3.10/site-packages/accelerate/commands/accelerate_cli.py", line 48, in main
    args.func(args)
  File "/home/ubuntu/SimpleTuner/.venv/lib/python3.10/site-packages/accelerate/commands/launch.py", line 1155, in launch_command
    multi_gpu_launcher(args)
  File "/home/ubuntu/SimpleTuner/.venv/lib/python3.10/site-packages/accelerate/commands/launch.py", line 793, in multi_gpu_launcher
    distrib_run.run(args)
  File "/home/ubuntu/SimpleTuner/.venv/lib/python3.10/site-packages/torch/distributed/run.py", line 910, in run
    elastic_launch(
  File "/home/ubuntu/SimpleTuner/.venv/lib/python3.10/site-packages/torch/distributed/launcher/api.py", line 138, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
  File "/home/ubuntu/SimpleTuner/.venv/lib/python3.10/site-packages/torch/distributed/launcher/api.py", line 260, in launch_agent
    result = agent.run()
  File "/home/ubuntu/SimpleTuner/.venv/lib/python3.10/site-packages/torch/distributed/elastic/metrics/api.py", line 137, in wrapper
    result = f(*args, **kwargs)
  File "/home/ubuntu/SimpleTuner/.venv/lib/python3.10/site-packages/torch/distributed/elastic/agent/server/api.py", line 710, in run
    self._shutdown()
  File "/home/ubuntu/SimpleTuner/.venv/lib/python3.10/site-packages/torch/distributed/elastic/agent/server/local_elastic_agent.py", line 365, in _shutdown
    self._pcontext.close(death_sig)
  File "/home/ubuntu/SimpleTuner/.venv/lib/python3.10/site-packages/torch/distributed/elastic/multiprocessing/api.py", line 572, in close
    self._close(death_sig=death_sig, timeout=timeout)
  File "/home/ubuntu/SimpleTuner/.venv/lib/python3.10/site-packages/torch/distributed/elastic/multiprocessing/api.py", line 909, in _close
    handler.proc.wait(time_to_wait)
  File "/usr/lib/python3.10/subprocess.py", line 1209, in wait
    return self._wait(timeout=timeout)
  File "/usr/lib/python3.10/subprocess.py", line 1953, in _wait
    time.sleep(delay)
  File "/home/ubuntu/SimpleTuner/.venv/lib/python3.10/site-packages/torch/distributed/elastic/multiprocessing/api.py", line 84, in _terminate_process_handler
    raise SignalException(f"Process {os.getpid()} got signal: {sigval}", sigval=sigval)
torch.distributed.elastic.multiprocessing.api.SignalException: Process 2006575 got signal: 2

**Reply by bghira** [reply]
if you want to use FSDP i would just rely on checkpoints instead of the final model export. see if that works.

**Reply by bghira** [reply]
i think the deepspeed zero3 was definitely sharding things because when you enable it, the CLIP and T5 models start sharding and this caused the earlier bugs.

the transformer model is sharded or it wouldn't have been able to even create the optimiser states.

**Reply by mobilejammer** [reply]
> if you want to use FSDP i would just rely on checkpoints instead of the final model export. see if that works.
When the Fully Sharded Data Parallel (FSDP) distributed training mode reaches the checkpoint, it gets stuck and doesn't continue to execute. There is no model in the output directory.

**Reply by bghira** [reply]
i guess that would have to be examined and fixed first, and i do not have the time

**Comment by mobilejammer** [comment]
one of my deepspeed conifg json file is as follow:
{
    "train_micro_batch_size_per_gpu": 1,
    "train_batch_size": 8,
    "gradient_clipping": 1.0,
    "gradient_checkpointing": true,
    "zero_optimization": {
        "stage": 3,
        "offload_optimizer": {
            "device": "nvme",
            "nvme_path": "/data/deepspeed_offload",
            "pin_memory": true,
            "buffer_count": 500
        },
        "offload_param": {
            "device": "cpu"
        },
        "overlap_comm": true,
        "contiguous_gradients": false,
        "sub_group_size": 5e6,
        "reduce_bucket_size": 5e3,
        "stage3_prefetch_bucket_size": 5e3,
        "stage3_param_persistence_threshold": 1e6,
        "stage3_max_live_parameters": 1e4,
        "stage3_max_reuse_distance": 1e3,
        "stage3_gather_16bit_weights_on_model_save": true
    },
    "pipeline": {
        "activation_checkpointing": {
            "partition_activations": true,
            "cpu_checkpointing": false,
            "checkpoint_in_cpu_mem": false
        },
        "num_stages": 8,
        "stage_gather_with_gradients": false
    },
    "aio": {
            "block_size": 5168576,
            "queue_depth": 8,
            "thread_count": 4,
            "single_submit": true,
            "overlap_events": true
        }
}
No matter how I set the "num_stages" parameter insight  config above, the eight 4090 graphics cards are always fully occupied and an "out of memory" error occurs. Logically speaking, it should be sufficient to perform full-parameter fine-tuning on flux-dev with eight cards. When the pipeline is not set, the graphics cards are also fully occupied and then the "out of memory" issue arises. So I somewhat suspect whether the graphic memory is allocated by zero3 is really working.

**Reply by bghira** [reply]
i don't know how you decided whether a certain number of GPUs should be sufficient but if it's not working, it's not a bug in this project. it would be a bug in DeepSpeed. and reporting issues there and having them investigated is a rather involved process I do not have time for; but I also don't have time to try and reproduce this problem beyond what I've already done to ensure Zero 3 is working. i have generally bad news for you that ZeRO3 ends up disabling validations and other things just to progress to the training loop. the 4090 probably just doesn't have enough memory. use L40S.

**Reply by bghira** [reply]
the problem may actually be that you're using some deepspeed config json instead of `accelerate config` as outlined in the DEEPSPEED.md document in this repo's documentation folder. using deepspeed config json is flaky and less likely to work. there's some caveats in Huggingface documentation about this, but I'm not sure where, offhand.

**Reply by mobilejammer** [reply]
> the problem may actually be that you're using some deepspeed config json instead of `accelerate config` as outlined in the DEEPSPEED.md document in this repo's documentation folder. using deepspeed config json is flaky and less likely to work. there's some caveats in Huggingface documentation about this, but I'm not sure where, offhand.

Thank you! At first, I also configured the accelerate config according to your readme.md. Because I thought that for the Flux model with 12 billion parameters, 24GB of memory(nvidia 4090) would be sufficient. And if it was sharded, it would be even more than enough to distribute it across 8 graphics cards. Later, I found that it didn't work, so I configured the json file, but it still didn't work. Then I tested the pipeline and it didn't work either. So, even without sharding on low-end graphics cards, there's a high probability that it's due to my operations and misunderstandings. I'll put this project aside for now and try other projects instead. Wish me good luck.

**Reply by bghira** [reply]
yes, the pipeline won't really work with ZERO3, this is a bug in Diffusers but it's notably difficult to resolve it, and so they do not.
